fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,176 @@
+        };
+    }
+
+    <T> Function<? super FDBRecordContext, CompletableFuture<? extends T>> workInSessionAsync(
+            @Nonnull Function<? super FDBRecordContext, CompletableFuture<? extends T>> work) {
+        return context -> checkLock(context)
+                .thenCompose(vignore -> work.apply(context))
+                .thenApply(result -> {
+                    setLockSessionTime(context.ensureActive());
+                    return result;
+                });
+    }
+
+    @Nonnull
+    UUID getSessionId() {
+        return sessionId;
+    }
+
+    private CompletableFuture<Void> checkLock(FDBRecordContext context) {
+        return getLockSessionId(context.ensureActive())
+                .thenCompose(lockSessionId -> {
+                    if (!sessionId.equals(lockSessionId)) { // sessionId is nonnull while lockSessionId is nullable
+                        throw new SynchronizedSessionExpiredException("Failed to continue the session")
+                                .addLogInfo(LogMessageKeys.SUBSPACE_KEY, ByteArrayUtil2.loggable(lockSubspace.getKey()))
+                                .addLogInfo(LogMessageKeys.UUID, sessionId)
+                                .addLogInfo("lockSessionId", lockSessionId);
+                    }
+                    return AsyncUtil.DONE;
+                });
+    }
+
+    void close(FDBRecordContext context) {
+        context.ensureActive().clear(lockSubspace.pack(), ByteArrayUtil.strinc(lockSubspace.pack()));
+    }
+
+    private CompletableFuture<UUID> getLockSessionId(@Nonnull Transaction tr) {
+        return tr.get(lockSessionIdSubspaceKey)
+                .thenApply(value -> value == null ? null : Tuple.fromBytes(value).getUUID(0));
+    }
+
+    private void setLockSessionId(@Nonnull Transaction tr) {
+        tr.set(lockSessionIdSubspaceKey, Tuple.from(sessionId).pack());
+    }
+
+    // There may be multiple threads working in a same session, in which case the session time is being written
+    // frequently. To avoid unnecessary races:
+    // - The session time should not be read while working in the session, so that all work transactions can write to
+    //   it blindly and not conflict with each other.
+    // - The session time should be updated to the max value when being updated by concurrent transactions, so that the
+    //   final value comes not from whoever gets committed last but whoever writes the largest value
+    // - When the session time is read during session initialization, it should be a snapshot read so it will not have
+    //   conflicts with working transactions (which write to session time).
+    private CompletableFuture<Long> getLockSessionTime(@Nonnull ReadTransaction tr) {
+        return tr.get(lockSessionTimeSubspaceKey)
+                .thenApply(value -> Tuple.fromBytes(value).getLong(0));
+    }
+
+    private void setLockSessionTime(@Nonnull Transaction tr) {
+        long leaseEndTime = System.currentTimeMillis() + leaseLengthMill;
```

Comment:
I'm getting ambiguous information online as to how Java's `System.currentTimeMillis` function handles leap seconds. I think it's supposed to skew them evenly throughout the day, which would be nice. I think because the lock UUID is the source of truth for whether the lock is held and the time is just being used here as a heuristic to see if the lock should be considered dead by others that it doesn't quite matter, as it just means that the lock will be held for 1 second longer than expected (if we leap backwards) or 1 second shorter than expected (if we leap forwards--which has never happened) and the leap second isn't skewed. Does that sound right?

Reply:
Yeah, I think you're right that clock drift is more worrisome than the leap second thing, but yeah, because we're only using time as a heuristic, and the source of truth is the lock session UUID, this should be fine.